party-manager.js

```diff
@@ -0,0 +1,163 @@
+import * as THREE from 'three';
+import physicsManager from './physics-manager.js';
+import {world} from './world.js';
+import {playersManager} from './players-manager.js';
+
+const localVector = new THREE.Vector3();
+
+const physicsScene = physicsManager.getScene();
+
+class PartyManager extends EventTarget {
+  constructor() {
+    super();
+    
+    this.partyPlayers = [];
+    this.targetMap = new Map();
+    this.onFrameMap = new Map();
+    this.onActivateMap = new Map();
+  }
+
+  destroy() {
+    for (const player of this.partyPlayers) {
+      this.#removePartyPlayer(player);
+    }
+  }
+
+  // switch to next character and returns main character
+  switchCharacter() {
+    const localPlayer = playersManager.getLocalPlayer();
+    
+    if (this.partyPlayers.length != 0) {
+      const nextPlayer = this.partyPlayers.shift();
+
+      localPlayer.isLocalPlayer = false;
+      localPlayer.isNpcPlayer = true;
+
+      nextPlayer.isLocalPlayer = true;
+      nextPlayer.isNpcPlayer = false;
+
+      nextPlayer.updatePhysicsStatus();
+      localPlayer.updatePhysicsStatus();
+
+      this.#setFollowTarget(nextPlayer, null);
+
+      this.partyPlayers.push(localPlayer);
+
+      this.#queueFollow(nextPlayer);
+
+      return nextPlayer;
+    }
+    return null;
+  }
+
+  // add new player to party
+  addPartyPlayer(newPlayer) {
+    const localPlayer = playersManager.getLocalPlayer();
+
+    if (this.partyPlayers.length < 2) { // 3 max members
+        this.#setFollowTarget(newPlayer, localPlayer);
+        this.partyPlayers.push(newPlayer);
+        newPlayer.isNpcInParty = true;
+    
+        this.#queueFollow(localPlayer);
+    
+        return true;  
+    }
+    return false;
+  }
+
+  #removePartyPlayer(player) {
+    const playerIndex = this.partyPlayers.indexOf(player);
+    if (!player.isMainPlayer && playerIndex !== -1) {
+      this.#setFollowTarget(player, null);
+      this.partyPlayers.splice(playerIndex, 1);
+      player.isNpcInParty = false;
+
+      const localPlayer = playersManager.getLocalPlayer();
+      this.#queueFollow(localPlayer);
+
+      return true;
+    }
+    return false;
+  }
+
+  // queue all party members to follow main player in a line
+  #queueFollow(mainPlayer) {
+    let headPlayer = mainPlayer;
+    for(const partyPlayer of this.partyPlayers) {
+      this.#setFollowTarget(partyPlayer, headPlayer);
+      headPlayer = partyPlayer;
+    }
+  }
+
+  #getPlayerApp(player) {
+    let avatarApp = player.avatar.app;
```

Comment:
I don't think we should reach in and look at the app like this -- in fact in the avatar optimizations branch this is probably impossible, as the app is being dynamically re-optimized under the avatar.

If we need to listen for events, the player/avatar can route the events from the underlying app.